File: mca-language-server/server/test/test_validator.js
const assert = require('assert');
const { validateDocument } = require('../src/validator');

function test(name, fn) {
  try {
    fn();
    console.log(`  PASS  ${name}`);
  } catch (err) {
    console.error(`  FAIL  ${name}`);
    console.error(`        ${err.message}`);
    process.exitCode = 1;
  }
}

// Helper: find diagnostics by code
function findByCode(diags, code) {
  return diags.filter(d => d.code === code);
}
function countByCode(diags, code) {
  return findByCode(diags, code).length;
}

// ── Colon tests ────────────────────────────────────────────────────────

test('missing_colon - ERROR', () => {
  const diags = validateDocument('no colon here');
  assert.strictEqual(countByCode(diags, 'missing_colon'), 1);
});

test('space_after_colon - WARNING', () => {
  const diags = validateDocument('test: /say hi');
  assert.strictEqual(countByCode(diags, 'space_after_colon'), 1);
});

test('no_content_after_colon - ERROR', () => {
  const diags = validateDocument('test:');
  assert.strictEqual(countByCode(diags, 'no_content_after_colon'), 1);
});

test('trailing_colons - ERROR', () => {
  const diags = validateDocument('test:/say:');
  assert.strictEqual(countByCode(diags, 'trailing_colons'), 1);
});

test('trailing_semicolon - INFO', () => {
  const diags = validateDocument('test:/say;');
  assert.strictEqual(countByCode(diags, 'trailing_semicolon'), 1);
});

// ── Bracket tests ──────────────────────────────────────────────────────

test('balanced_square_brackets - no error', () => {
  const diags = validateDocument('test:/say [test]');
  assert.strictEqual(countByCode(diags, 'unbalanced_square'), 0);
  assert.strictEqual(countByCode(diags, 'unmatched_close_square'), 0);
});

test('unmatched_close_square - ERROR', () => {
  const diags = validateDocument('test:/say ]');
  assert.strictEqual(countByCode(diags, 'unmatched_close_square'), 1);
});

test('unbalanced_open_square - ERROR', () => {
  const diags = validateDocument('test:/say [test');
  assert.strictEqual(countByCode(diags, 'unbalanced_square'), 1);
});

test('balanced_curly_brackets - no error', () => {
  const diags = validateDocument('test:/say {nbt}');
  assert.strictEqual(countByCode(diags, 'unbalanced_curly'), 0);
  assert.strictEqual(countByCode(diags, 'unmatched_close_curly'), 0);
});

test('unmatched_close_curly - ERROR', () => {
  const diags = validateDocument('test:/say }');
  assert.strictEqual(countByCode(diags, 'unmatched_close_curly'), 1);
});

test('unbalanced_open_curly - ERROR', () => {
  const diags = validateDocument('test:/say {nbt');
  assert.strictEqual(countByCode(diags, 'unbalanced_curly'), 1);
});

test('brackets_inside_double_quotes', () => {
  const diags = validateDocument('test:/say "hello [world]"');
  assert.strictEqual(countByCode(diags, 'unbalanced_square'), 0);
});

test('brackets_inside_single_quotes', () => {
  const diags = validateDocument("test:/say 'hello [world]'");
  assert.strictEqual(countByCode(diags, 'unbalanced_square'), 0);
});

test('escaped_quotes_inside_brackets', () => {
  const diags = validateDocument("test:/say [it\\'s ok]");
  assert.strictEqual(countByCode(diags, 'unbalanced_square'), 0);
});

// ── Trigger name tests ─────────────────────────────────────────────────

test('valid trigger name', () => {
  const diags = validateDocument('like:/say Thanks!');
  assert.strictEqual(countByCode(diags, 'invalid_trigger_name'), 0);
});

test('valid quoted trigger name', () => {
  const diags = validateDocument("'my trigger':/say hi");
  assert.strictEqual(countByCode(diags, 'invalid_trigger_name'), 0);
});

test('invalid trigger name (special chars)', () => {
  const diags = validateDocument('bad-trigger!:x');
  assert.strictEqual(countByCode(diags, 'invalid_trigger_name'), 1);
});

test('invalid quoted trigger name', () => {
  const diags = validateDocument("'bad-trigger!':x");
  assert.strictEqual(countByCode(diags, 'invalid_trigger_name'), 1);
});

test('duplicate_trigger', () => {
  const diags = validateDocument('dup:/a\ndup:/b');
  assert.strictEqual(countByCode(diags, 'duplicate_trigger'), 1);
});

// ── Command prefix tests ───────────────────────────────────────────────

test('valid slash prefix', () => {
  const diags = validateDocument('test:/say hi');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 0);
});

test('valid dollar prefix', () => {
  const diags = validateDocument('test:$script');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 0);
});

test('valid bang prefix', () => {
  const diags = validateDocument('test:!rcon cmd');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 0);
});

test('valid overlay arrow', () => {
  const diags = validateDocument('test:>>overlay');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 0);
});

test('valid named overlay', () => {
  const diags = validateDocument('test:@name>>overlay');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 0);
});

test('valid ampersand prefix', () => {
  const diags = validateDocument('test:&curl http://localhost');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 0);
});

test('invalid prefix', () => {
  const diags = validateDocument('test:%bad');
  assert.strictEqual(countByCode(diags, 'invalid_prefix'), 1);
});

// ── Placeholder tests ──────────────────────────────────────────────────

test('{comment} on comment trigger - no error', () => {
  const diags = validateDocument('comment:>>say {comment}');
  assert.strictEqual(countByCode(diags, 'comment_placeholder_wrong_trigger'), 0);
});

test('{comment} on non-comment trigger - ERROR', () => {
  const diags = validateDocument('like:>>say {comment}');
  assert.strictEqual(countByCode(diags, 'comment_placeholder_wrong_trigger'), 1);
});

// ── Multiplier tests ───────────────────────────────────────────────────

test('valid multiplier', () => {
  const diags = validateDocument('test:/cmd x5');
  assert.strictEqual(countByCode(diags, 'invalid_multiplier'), 0);
});

test('high multiplier warning', () => {
  const diags = validateDocument('test:/cmd x100');
  assert.strictEqual(countByCode(diags, 'high_multi'), 1);
});

test('high multiplier suppressed with ignore-lag', () => {
  const diags = validateDocument('test:/cmd x100 # ignore-lag');
  assert.strictEqual(countByCode(diags, 'high_multi'), 0);
});

test('invalid multiplier (non-numeric)', () => {
  const diags = validateDocument('test:/cmd xabc');
  assert.strictEqual(countByCode(diags, 'invalid_multiplier'), 1);
});

test('overlay multiplier not allowed', () => {
  const diags = validateDocument('test:>>overlay x5');
  assert.strictEqual(countByCode(diags, 'overlay_multiplier'), 1);
});

// ── Empty lines and comments ───────────────────────────────────────────

test('empty lines produce no diagnostics', () => {
  const diags = validateDocument('\n\n');
  assert.strictEqual(diags.length, 0);
});

test('full-line comment produces no diagnostics', () => {
  const diags = validateDocument('# this is a comment');
  assert.strictEqual(diags.length, 0);
});

test('inline comment is valid', () => {
  const diags = validateDocument('test:/say hi # inline comment');
  const errors = diags.filter(d => d.severity === 1);
  assert.strictEqual(errors.length, 0);
});

// ── Semicolon separated commands ───────────────────────────────────────

test('multiple commands - valid', () => {
  const diags = validateDocument('test:/first;/second');
  const errors = diags.filter(d => d.severity === 1);
  assert.strictEqual(errors.length, 0);
});

test('empty command block (double semicolon)', () => {
  const diags = validateDocument('test:/first;;/third');
  assert.strictEqual(countByCode(diags, 'empty_command_block'), 1);
});

test('trailing semicolon after multiple commands', () => {
  const diags = validateDocument('test:/first;/second;');
  assert.strictEqual(countByCode(diags, 'trailing_semicolon'), 1);
});

// ── Disabled triggers ──────────────────────────────────────────────────

test('disabled trigger ## is validated', () => {
  const diags = validateDocument('##disabled:!cmd');
  const errors = diags.filter(d => d.severity === 1);
  assert.strictEqual(errors.length, 0); // valid disabled trigger
});

test('disabled trigger with ## and duplicate', () => {
  const diags = validateDocument('##dup:/a\ndup:/b');
  assert.strictEqual(countByCode(diags, 'duplicate_trigger'), 1);
});

// ── Robustness ─────────────────────────────────────────────────────────

test('malformed file does not crash', () => {
  const diags = validateDocument(':\n  \n:::');
  assert.ok(Array.isArray(diags));
});

test('incomplete line while typing', () => {
  const diags = validateDocument('test:');
  assert.strictEqual(countByCode(diags, 'no_content_after_colon'), 1);
});

test('empty document', () => {
  const diags = validateDocument('');
  assert.strictEqual(diags.length, 0);
});

// ── Real-world examples ────────────────────────────────────────────────

test('default actions.mca has no errors', () => {
  const text = (
    "follow:/give @a minecraft:golden_apple 7 ; >>New Follower!|{user} is now following you!|5\n" +
    "likes:/execute at @a run summon minecraft:creeper ~ ~ ~ x2\n" +
    "like_2:/clear @a * ; /kill @a\n" +
    "5655:!tnt 2 0.1 2 Notch\n" +
    "16111:/give @a minecraft:diamond\n" +
    "8913:/execute at @a run summon minecraft:evoker ~ ~ ~ x3\n" +
    "6267:!tnt 600 0.1 2 Notch\n" +
    "7168:/execute at @a run summon minecraft:wither ~ ~ ~ x50\n" +
    "16071:$random\n" +
    "'Tom the Tomato':$random\n" +
    "16379:$random\n"
  );
  const diags = validateDocument(text);
  const errors = diags.filter(d => d.severity === 1);
  assert.strictEqual(errors.length, 0, `Unexpected errors: ${JSON.stringify(errors)}`);
});

console.log('\nValidator tests complete.\n');

File: tests/test_core/test_validator.py
import pytest

pytestmark = pytest.mark.validator


class TestValidatorBrackets:
    def test_balanced_square_brackets(self):
        from core.validator import validate_text

        diags = validate_text("test:/say [test]")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_unmatched_close_square(self):
        from core.validator import validate_text

        diags = validate_text("test:/say ]")
        errors = [d for d in diags if d.code == "unmatched_close_square"]
        assert len(errors) == 1

    def test_unbalanced_open_square(self):
        from core.validator import validate_text

        diags = validate_text("test:/say [test")
        errors = [d for d in diags if d.code == "unbalanced_square"]
        assert len(errors) == 1

    def test_balanced_curly_brackets(self):
        from core.validator import validate_text

        diags = validate_text("test:/say {nbt}")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_unmatched_close_curly(self):
        from core.validator import validate_text

        diags = validate_text("test:/say }")
        errors = [d for d in diags if d.code == "unmatched_close_curly"]
        assert len(errors) == 1

    def test_unbalanced_open_curly(self):
        from core.validator import validate_text

        diags = validate_text("test:/say {nbt")
        errors = [d for d in diags if d.code == "unbalanced_curly"]
        assert len(errors) == 1

    def test_brackets_inside_quotes(self):
        from core.validator import validate_text

        diags = validate_text('test:/say "hello [world]"')
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_brackets_inside_single_quotes(self):
        from core.validator import validate_text

        diags = validate_text("test:/say 'hello [world]'")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_escaped_quotes_inside_brackets(self):
        from core.validator import validate_text

        diags = validate_text(r"test:/say [it\'s ok]")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0


class TestValidatorColons:
    def test_missing_colon(self):
        from core.validator import validate_text

        diags = validate_text("no colon here")
        errors = [d for d in diags if d.code == "missing_colon"]
        assert len(errors) == 1

    def test_space_after_colon(self):
        from core.validator import validate_text

        diags = validate_text("test: /say hi")
        warnings = [d for d in diags if d.code == "space_after_colon"]
        assert len(warnings) == 1

    def test_no_content_after_colon(self):
        from core.validator import validate_text

        diags = validate_text("test:")
        errors = [d for d in diags if d.code == "no_content_after_colon"]
        assert len(errors) == 1

    def test_trailing_colon(self):
        from core.validator import validate_text

        diags = validate_text("test:/say:")
        errors = [d for d in diags if d.code == "trailing_colons"]
        assert len(errors) == 1

    def test_trailing_semicolon(self):
        from core.validator import validate_text

        diags = validate_text("test:/say;")
        infos = [d for d in diags if d.code == "trailing_semicolon"]
        assert len(infos) == 1


class TestValidatorTriggerNames:
    def test_valid_trigger(self):
        from core.validator import validate_text

        diags = validate_text("like:/say Thanks!")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_valid_quoted_trigger(self):
        from core.validator import validate_text

        diags = validate_text("'my trigger':/say hi")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_invalid_trigger_special_chars(self):
        from core.validator import validate_text

        diags = validate_text("bad-trigger!:x")
        errors = [d for d in diags if d.code == "invalid_trigger_name"]
        assert len(errors) == 1

    def test_invalid_quoted_trigger(self):
        from core.validator import validate_text

        diags = validate_text("'bad-trigger!':x")
        errors = [d for d in diags if d.code == "invalid_trigger_name"]
        assert len(errors) == 1

    def test_duplicate_trigger(self):
        from core.validator import validate_text

        diags = validate_text("dup:/a\ndup:/b")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 1

    def test_disabled_trigger_not_counted_as_duplicate(self):
        from core.validator import validate_text

        diags = validate_text("##dup:/a\ndup:/b")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 0

    def test_disabled_trigger_colliding_with_active_warns(self):
        from core.validator import validate_text

        diags = validate_text("like_2:/kill @a\n##like_2:/kill @a")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 0
        warnings = [d for d in diags if d.code == "duplicate_trigger_disabled"]
        assert len(warnings) == 1
        assert warnings[0].severity.name == "WARNING"
        assert warnings[0].line == 1

    def test_active_trigger_colliding_with_earlier_disabled_warns(self):
        from core.validator import validate_text

        diags = validate_text("##dup:/a\ndup:/b")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 0
        warnings = [d for d in diags if d.code == "duplicate_trigger_disabled"]
        assert len(warnings) == 1


class TestValidatorDisabledTriggers:
    def test_disabled_duplicate_pair_not_flagged(self):
        """Two disabled (##) triggers with the same name are OFF and must
        not count as duplicates."""
        from core.validator import validate_text

        diags = validate_text("##dup:/a\n##dup:/b")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 0

    def test_disabled_trigger_does_not_duplicate_active(self):
        """A disabled (##) trigger sharing a name with an active trigger is
        OFF and must not be reported as a duplicate."""
        from core.validator import validate_text

        diags = validate_text("dup:/a\n##dup:/b")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 0

    def test_active_duplicate_still_flagged(self):
        """Two active triggers with the same name are still errors."""
        from core.validator import validate_text

        diags = validate_text("dup:/a\ndup:/b")
        errors = [d for d in diags if d.code == "duplicate_trigger"]
        assert len(errors) == 1

    def test_disabled_trigger_still_syntax_validated(self):
        """Disabled triggers are kept as templates; their content is still
        validated for syntax errors."""
        from core.validator import validate_text

        diags = validate_text("##nocolonhere")
        errors = [d for d in diags if d.code == "missing_colon"]
        assert len(errors) == 1


class TestValidatorCommandPrefixes:
    def test_valid_slash(self):
        from core.validator import validate_text

        diags = validate_text("test:/say hi")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_valid_dollar(self):
        from core.validator import validate_text

        diags = validate_text("test:$script")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_valid_bang(self):
        from core.validator import validate_text

        diags = validate_text("test:!rcon cmd")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_valid_overlay_arrow(self):
        from core.validator import validate_text

        diags = validate_text("test:>>overlay")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_valid_overlay_named(self):
        from core.validator import validate_text

        diags = validate_text("test:@name>>overlay")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_valid_ampersand(self):
        from core.validator import validate_text

        diags = validate_text("test:&curl http://localhost")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_invalid_prefix(self):
        from core.validator import validate_text

        diags = validate_text("test:%bad")
        errors = [d for d in diags if d.code == "invalid_prefix"]
        assert len(errors) == 1


class TestValidatorPlaceholder:
    def test_comment_placeholder_on_comment_trigger(self):
        from core.validator import validate_text

        diags = validate_text("comment:>>say {comment}")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_comment_placeholder_wrong_trigger(self):
        from core.validator import validate_text

        diags = validate_text("like:>>say {comment}")
        errors = [d for d in diags if d.code == "comment_placeholder_wrong_trigger"]
        assert len(errors) == 1


class TestValidatorMultipliers:
    def test_valid_multiplier(self):
        from core.validator import validate_text

        diags = validate_text("test:/cmd x5")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_high_multiplier_warning(self):
        from core.validator import validate_text

        diags = validate_text("test:/cmd x100")
        warnings = [d for d in diags if d.code == "high_multi"]
        assert len(warnings) == 1

    def test_high_multiplier_suppressed_with_ignore_lag(self):
        from core.validator import validate_text

        diags = validate_text("test:/cmd x100 # ignore-lag")
        warnings = [d for d in diags if d.code == "high_multi"]
        assert len(warnings) == 0

    def test_invalid_multiplier(self):
        from core.validator import validate_text

        diags = validate_text("test:/cmd xabc")
        errors = [d for d in diags if d.code == "invalid_multiplier"]
        assert len(errors) == 1

    def test_overlay_multiplier_not_allowed(self):
        from core.validator import validate_text

        diags = validate_text("test:>>overlay x5")
        errors = [d for d in diags if d.code == "overlay_multiplier"]
        assert len(errors) == 1


class TestValidatorEmptyAndComments:
    def test_empty_line_skipped(self):
        from core.validator import validate_text

        diags = validate_text("\n\n")
        assert len(diags) == 0

    def test_comment_line_skipped(self):
        from core.validator import validate_text

        diags = validate_text("# this is a comment")
        assert len(diags) == 0

    def test_inline_comment(self):
        from core.validator import validate_text

        diags = validate_text("test:/say hi # inline comment")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0


class TestValidatorSemicolonSeparatedCommands:
    def test_multiple_commands(self):
        from core.validator import validate_text

        diags = validate_text("test:/first;/second")
        errors = [d for d in diags if d.severity.name == "ERROR"]
        assert len(errors) == 0

    def test_empty_command_block(self):
        from core.validator import validate_text

        diags = validate_text("test:/first;;/third")
        warnings = [d for d in diags if d.code == "empty_command_block"]
        assert len(warnings) == 1

    def test_trailing_semicolon_after_multiple(self):
        from core.validator import validate_text

        diags = validate_text("test:/first;/second;")
        infos = [d for d in diags if d.code == "trailing_semicolon"]
        assert len(infos) == 1


class TestValidatorFileLevel:
    def test_validate_file_not_found(self):
        from core.validator import validate_file

        with pytest.raises(FileNotFoundError):
            validate_file("/nonexistent/path.mca")

    def test_validate_file_valid(self, tmp_path):
        from core.validator import validate_file

        f = tmp_path / "valid.mca"
        f.write_text("test:/say hi", encoding="utf-8")
        diags = validate_file(f, raise_on_error=False)
        assert len(diags) == 0

    def test_validate_file_raises_on_error(self, tmp_path):
        from core.validator import validate_file

        f = tmp_path / "bad.mca"
        f.write_text("no colon", encoding="utf-8")
        with pytest.raises(ValueError, match="Validation failed"):
            validate_file(f, raise_on_error=True)

    def test_validate_file_no_raise(self, tmp_path):
        from core.validator import validate_file

        f = tmp_path / "bad.mca"
        f.write_text("no colon", encoding="utf-8")
        diags = validate_file(f, raise_on_error=False)
        assert len(diags) == 1
        assert diags[0].code == "missing_colon"


class TestValidatorSeverity:
    def test_error_severity(self):
        from core.validator import Severity

        assert Severity.ERROR.value == "ERROR"
        assert Severity.WARNING.value == "WARNING"
        assert Severity.INFO.value == "INFO"

    def test_diagnostic_fields(self):
        from core.validator import Diagnostic, Severity

        d = Diagnostic(
            line=0,
            start_char=0,
            end_char=5,
            message="test",
            severity=Severity.ERROR,
            code="test_code",
        )
        assert d.line == 0
        assert d.start_char == 0
        assert d.end_char == 5
        assert d.message == "test"
        assert d.severity == Severity.ERROR
        assert d.code == "test_code"
